全部屋のスコアが未入力の場合、お部屋のスコア結果ページに遷移できないようにする

uchihiro04 opened the pull request:
## 概要 
全部屋のスコアが未入力の場合、以下のような形になるように変更しました。
* 「スコアを見る」をタップ・クリックしてもスコア結果ページに遷移しない。
* お部屋のスコア結果ページに直接アクセスすると内見一覧ページにリダイレクトする。
* リダイレクト時は「スコアが未入力です」というフラッシュメッセージが表示される。

## スクリーンショット
* 全部屋のスコアが未入力の状態で、「スコアを見る」ボタンのクリックを試みた場合（リンクのない状態）

https://github.com/uchihiro04/house-viewing-score-log/assets/77523896/e92eeb74-95af-4bc8-a66c-8c32d2872691

* 全部屋のスコアが未入力の状態で、お部屋のスコア結果ページに直接アクセスした場合

https://github.com/uchihiro04/house-viewing-score-log/assets/77523896/8bbd6fa0-3ce5-4498-85ab-8d85c85c37b1

## 関連Issue
- close #80 

Close #80

uchihiro04 committed:
#80 「スコアを見る」ボタンに関する条件分岐を追加

内見一覧ページの「スコアを見る」ボタンについて、全部屋スコア未入力の場合はリンクが付かないように変更しました。

uchihiro04 committed:
#80 スコア未入力時にスコア結果ページに遷移した場合、内見一覧ページにリダイレクトする形に変更

uchihiro04 committed:
#80 フラッシュメッセージを設定

スコア未入力により内見一覧ページにリダイレクトした際、フラッシュメッセージを表示するよう設定しました。

Yuma-Nagaoka commented on app/views/layouts/application.html.slim:
```diff
@@ -11,5 +11,5 @@ html
   body
     main.container.mx-auto.mt-28.px-5
-      = flash.notice.presence
+      = flash.notice.presence || flash.alert.presence
```
ここnoticeがある時alertがあっても出ませんがそういう状況にはなりませんか？
もしくはそれで問題ないです？

uchihiro04 replied:
ご指摘ありがとうございます。
最初は以下のようにnoticeとalertを分けた作りにしていました。
```
= flash.notice.presence
= flash.alert.presence
```

しかし、noticeとalertが使用される場合を整理してみると以下の通り、両方が同時に存在する場面がないことが分かりました。
* noticeが適用される場面
  * スコアの入力・編集に成功し、内見一覧ページにリダイレクトした時
* alertが適用される場面
  * スコア未入力の状態でスコア結果ページ・スコア詳細ページにアクセスし、内見一覧ページにリダイレクトした時
  
そのため、現時点では`flash.notice.presence || flash.alert.presence`という形でも問題はなさそうだと考えています。
ただし、今後機能を拡張した時に分ける必要が出てくる可能性はあると思います。

Yuma-Nagaoka commented on app/helpers/house_viewings/rooms_helper.rb:
```diff
@@ -0,0 +1,9 @@
+module HouseViewings
+  module RoomsHelper
+    def scores?(rooms)
+      rooms.any? { |room| room.scores.exists? }
```
ここn+1のクエリが出てしまうので、
Score.where(room_id: room_ids).exists?
みたいな形にした方が良いです

uchihiro04 replied:
whereメソッドで、複数のカラムの数値を検索できることを初めて知りました。繰り返し処理を使うよりクエリの発行回数が少なくて済みますね……。
scores?メソッドについて、繰り返し処理ではなくwhereメソッドを使用する形に修正しました。該当するコミットは[ff12684](https://github.com/uchihiro04/house-viewing-score-log/pull/131/commits/ff126843cb20c440f688fad75be587000cf69f43)です。

uchihiro04 replied:
rubocopの指摘を受け、scores?メソッドについてwhereメソッドを間に挟まず、`Score.exists?(room_id: rooms.ids)`のような形に修正を行いました。該当するコミットは[9bcbd98](https://github.com/uchihiro04/house-viewing-score-log/pull/131/commits/9bcbd9837f3ec70b23eaf9103a32c4a4fd0f7d96)です。

uchihiro04 committed:
#80 scores?メソッドを修正

scores?メソッドについて、繰り返し処理を削除してクエリの実行回数が少なくなるように修正しました。

uchihiro04 committed:
#80 scores?メソッドからwhereメソッドを削除

rubocopの指摘を受けscores?メソッドについて、whereメソッドを挟まずに直接exists?メソッドを使用する形に変更しました。

takuminasu commented on app/controllers/house_viewings/scores_controller.rb:
```diff
@@ -8,7 +8,15 @@ def index
+
+    private
+
+    def rooms_score_blank?
+      @rooms.all? { |room| room.scores.blank? }
+    end
```
```
def scores?(rooms)
  Score.exists?(room_id: rooms.ids)
end
```

この2つメソッドがありますが空か存在するかを確認しているだけなので`Room.score_entered?(room_ids)`のようなメソッド1つで対応できると思います。

uchihiro04 replied:
ご指摘いただいた通り、行っている中身は2つともあまり変わらないため、1つのメソッドで対応できそうですね……。

いただいたレビューおよびハドルでのアドバイスをもとに、`rooms_score_blank?`メソッド、`scores?`メソッドを統合し、Roomモデルのクラスメソッド（`score_entered?`）として定義しました。該当するコミットは[8b423cf](https://github.com/uchihiro04/house-viewing-score-log/pull/131/commits/8b423cf9f4b30304829006e069e08af3af73ec4f)です。

uchihiro04 committed:
#80 メソッドの統合

ヘルパーメソッドのscores?メソッドとHouseViewings::Scoresコントローラのrooms_score_blank?メソッドを統合し、Roomモデルのクラスメソッド（score_entered?）として定義しました。

takuminasu reviewed (approved):
LGTM
